File: js/lib/VisVersion.ts
export enum VisVersion {
    v3_4a = "3-4a",
    v3_5a = "3-5a",
    v3_6a = "3-6a",
    v3_7a = "3-7a",
}

export class VisVersionExtension {
    public static toVersionString(
        version: VisVersion,
        builder?: string[]
    ): string {
        let v: string;
        switch (version) {
            case VisVersion.v3_4a:
                v = "3-4a";
                break;
            case VisVersion.v3_5a:
                v = "3-5a";
                break;
            case VisVersion.v3_6a:
                v = "3-6a";
                break;
            case VisVersion.v3_7a:
                v = "3-7a";
                break;
            default:
                throw new Error("Invalid VisVersion enum value: " + version);
        }

        if (builder) {
            builder.push(v);
        }
        return v;
    }

    public static toString(version: VisVersion, builder?: string[]): string {
        let v: string;
        switch (version) {
            case VisVersion.v3_4a:
                v = "3-4a";
                break;
            case VisVersion.v3_5a:
                v = "3-5a";
                break;
            case VisVersion.v3_6a:
                v = "3-6a";
                break;
            case VisVersion.v3_7a:
                v = "3-7a";
                break;
            default:
                throw new Error("Invalid VisVersion enum value: " + version);
        }

        if (builder) {
            builder.push(v);
        }
        return v;
    }

    public static isValid(version: VisVersion): boolean {
        switch (version) {
            case VisVersion.v3_4a:
                return true;
            case VisVersion.v3_5a:
                return true;
            case VisVersion.v3_6a:
                return true;
            case VisVersion.v3_7a:
                return true;

            default:
                return false;
        }
    }
}

export class VisVersions {
    public static get all(): VisVersion[] {
        return Object.values(VisVersion)
            .map((v) => this.tryParse(v))
            .filter((v) => v) as VisVersion[];
    }

    public static parse(version: string): VisVersion {
        const v = this.tryParse(version);
        if (!v) {
            throw new Error("Couldnt parse version string: " + version);
        }

        return v;
    }

    public static tryParse(version: string): VisVersion | undefined {
        switch (version) {
            case "3-4a":
                return VisVersion.v3_4a;
            case "3-5a":
                return VisVersion.v3_5a;
            case "3-6a":
                return VisVersion.v3_6a;
            case "3-7a":
                return VisVersion.v3_7a;

            default:
                return;
        }
    }
}

File: cpp/src/dnv/vista/sdk/VISVersion.cpp
#include "pch.h"

#include "dnv/vista/sdk/VISVersion.h"

namespace dnv::vista::sdk
{
	bool VisVersionExtensions::tryParse( const std::string& versionString, VisVersion& version )
	{
		static const std::unordered_map<std::string_view, VisVersion> versionMap = {
			{ "3-4a", VisVersion::v3_4a },
			{ "3-5a", VisVersion::v3_5a },
			{ "3-6a", VisVersion::v3_6a },
			{ "3-7a", VisVersion::v3_7a },
			{ "3-8a", VisVersion::v3_8a },
			{ "3-9a", VisVersion::v3_9a },
		};

		auto it = versionMap.find( versionString );
		if ( it != versionMap.end() )
		{
			version = it->second;
			return true;
		}

		return false;
	}

	VisVersion VisVersionExtensions::parse( const std::string& versionString )
	{
		VisVersion version;
		if ( !tryParse( versionString, version ) )
		{
			throw std::invalid_argument( "Invalid VIS version string: " + versionString );
		}
		return version;
	}

	std::vector<VisVersion> VisVersionExtensions::allVersions()
	{
		static constexpr std::array<VisVersion, 6> versions{
			{ VisVersion::v3_4a,
				VisVersion::v3_5a,
				VisVersion::v3_6a,
				VisVersion::v3_7a,
				VisVersion::v3_8a,
				VisVersion::v3_9a } };

		return std::vector<VisVersion>( versions.begin(), versions.end() );
	}
}
